bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RowMergerV2.java

```diff
@@ -0,0 +1,462 @@
+      Preconditions.checkState(!hasChunkInProgess());
+      addCell(
+          Cell.newBuilder()
+              .setTimestampMicros(chunk.getTimestampMicros())
+              .addAllLabels(chunk.getLabelsList())
+              .setValue(chunk.getValue())
+              .build());
+    }
+
+    public void completeMultiChunkCell() {
+      Preconditions.checkArgument(hasChunkInProgess());
+      addCell(
+          Cell.newBuilder()
+              .setTimestampMicros(firstMultiChunkCell.getTimestampMicros())
+              .addAllLabels(firstMultiChunkCell.getLabelsList())
+              .setValue(BigtableZeroCopyByteStringUtil.wrap(outputStream.toByteArray()))
+              .build());
+      outputStream = null;
+      firstMultiChunkCell = null;
+    }
+
+    private void addCell(Cell cell) {
+      families.addCell(currentId.family, currentId.qualifier, cell);
+    }
+
+    /**
+     * update the current key with the new chunk info
+     */
+    void updateCurrentKey(ReadRowsResponse.CellChunk chunk) {
+      if (currentId == null || isNewRowKey(chunk)) {
+        currentId = new CellIdentifier(chunk);
+      } else if (chunk.hasFamilyName()) {
+        currentId = currentId.nextKeyForFamily(chunk);
+      } else if (chunk.hasQualifier()) {
+        currentId = currentId.nextKeyForQualifier(chunk);
+      } else {
+        currentId = currentId.nextKeyForTimestamp(chunk);
+      }
+    }
+
+    private boolean isNewRowKey(ReadRowsResponse.CellChunk chunk) {
+      return !chunk.getRowKey().isEmpty() && !chunk.getRowKey().equals(currentId.rowKey);
+    }
+
+    public boolean hasChunkInProgess() {
+      return outputStream != null;
+    }
+
+    void addPartialCellChunk(ReadRowsResponse.CellChunk chunk) throws IOException {
+      if (outputStream == null) {
+        outputStream = new ByteArrayOutputStream(chunk.getValueSize());
+        firstMultiChunkCell = chunk;
+      }
+      chunk.getValue().writeTo(outputStream);
+    }
+
+    public Row createRow() {
+      return Row.newBuilder().setKey(getRowKey()).addAllFamilies(families.getFamilies()).build();
+    }
+
+    public ByteString getRowKey() {
+      return currentId.rowKey;
+    }
+  }
+
+  private static boolean isCommit(CellChunk chunk) {
+    return chunk.getRowStatusCase() == RowStatusCase.COMMIT_ROW && chunk.getCommitRow();
+  }
+
+  private static boolean isReset(CellChunk chunk) {
+    return chunk.getRowStatusCase() == RowStatusCase.RESET_ROW && chunk.getResetRow();
+  }
+
+  private final StreamObserver<Row> observer;
+
+  private RowMergerState state = RowMergerState.NewRow;
+  private ByteString previousKey;
+  private RowInProgress rowInProgress;
+  private boolean complete;
+
+  public RowMergerV2(StreamObserver<Row> observer) {
+    this.observer = observer;
+  }
+
+  @Override
+  public void onNext(ReadRowsResponse partialRow) {
+    if (complete) {
+      observer.onError(new IllegalStateException("Adding partialRow after completion"));
```

Comment:
more consistent to use onError() even though you don't need complete to be set to true...